src/rust/inetstack/protocols/layer4/mod.rs

```diff
@@ -68,47 +84,64 @@ pub enum Socket<N: NetworkRuntime> {
 
-        Ok(Peer {
+        Ok(Layer4Endpoint {
             local_ipv4_addr: config.local_ipv4_addr()?,
             icmpv4,
             tcp,
             udp,
+            tcp_checksum_offload: config.tcp_checksum_offload()?,
+            udp_checksum_offload: config.udp_checksum_offload()?,
         })
     }
 
-    pub fn receive(&mut self, buf: DemiBuffer) {
-        let (header, payload) = match Ipv4Header::parse(buf) {
-            Ok(result) => result,
-            Err(e) => {
-                let cause: String = format!("Invalid destination address: {:?}", e);
-                warn!("dropping packet: {}", cause);
-                return;
+    pub fn receive(&mut self, ip_hdr: Ipv4Header, buf: DemiBuffer) -> Result<(), Fail> {
+        match ip_hdr.get_protocol() {
+            IpProtocol::TCP => {
+                let (tcp_hdr, data): (TcpHeader, DemiBuffer) =
+                    match TcpHeader::parse(&ip_hdr, buf, self.tcp_checksum_offload) {
+                        Ok(result) => result,
+                        Err(e) => {
+                            let cause: String = format!("invalid tcp header: {:?}", e);
+                            error!("receive(): {}", &cause);
+                            return Err(Fail::new(libc::EINVAL, &cause));
+                        },
+                    };
+                self.tcp
+                    .receive(ip_hdr.get_src_addr(), ip_hdr.get_dest_addr(), tcp_hdr, data);
             },
-        };
-        debug!("Ipv4 received {:?}", header);
-        if header.get_dest_addr() != self.local_ipv4_addr && !header.get_dest_addr().is_broadcast() {
-            let cause: String = format!("Invalid destination address");
-            warn!("dropping packet: {}", cause);
-            return;
-        }
-        match header.get_protocol() {
-            IpProtocol::ICMPv4 => self.icmpv4.receive(header, payload),
-            IpProtocol::TCP => self.tcp.receive(header, payload),
-            IpProtocol::UDP => self.udp.receive(header, payload),
+            IpProtocol::UDP => {
+                // Parse datagram.
+                let (udp_hdr, data): (UdpHeader, DemiBuffer) =
+                    match UdpHeader::parse(&ip_hdr, buf, self.udp_checksum_offload) {
+                        Ok(result) => result,
+                        Err(e) => {
+                            let cause: String = format!("dropping packet: unable to parse UDP header");
+                            warn!("{}: {:?}", cause, e);
+                            return Err(Fail::new(libc::EINVAL, &cause));
+                        },
+                    };
+                debug!("UDP received {:?}", udp_hdr);
+                self.udp
+                    .receive(ip_hdr.get_src_addr(), ip_hdr.get_dest_addr(), udp_hdr, data)
+            },
+            _ => unreachable!("Any other protocols should have been processed at the lower layer"),
         }
+        Ok(())
```

Comment:
Could we remove from this PR any changes that do not concern source tree refactoring?